src/main/java/io/cdap/plugin/gcp/gcs/sink/GCSOutputFormatProvider.java

```diff
@@ -0,0 +1,274 @@
+package io.cdap.plugin.gcp.gcs.sink;
+
+import com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem;
+import com.google.cloud.hadoop.gcsio.GoogleCloudStorageFileSystem;
+import com.google.cloud.hadoop.gcsio.StorageResourceId;
+import com.google.cloud.hadoop.gcsio.UpdatableItemInfo;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.etl.api.validation.FormatContext;
+import io.cdap.cdap.etl.api.validation.ValidatingOutputFormat;
+import io.cdap.plugin.gcp.gcs.GCSPath;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.JobStatus;
+import org.apache.hadoop.mapreduce.OutputCommitter;
+import org.apache.hadoop.mapreduce.OutputFormat;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter;
+import org.apache.hadoop.util.ReflectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * OutputFormatProvider for GCSSink
+ */
+public class GCSOutputFormatProvider implements ValidatingOutputFormat {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GCSOutputFormatProvider.class);
+  private static final String DELEGATE_OUTPUTFORMAT_CLASSNAME = "gcssink.delegate.outputformat.classname";
+  private static final String OUTPUT_FOLDER = "gcssink.metric.output.folder";
+  public static final String RECORD_COUNT_FORMAT = "recordcount.%s";
+  private ValidatingOutputFormat delegate;
+
+  public GCSOutputFormatProvider(ValidatingOutputFormat delegate) {
+    this.delegate = delegate;
+  }
+
+  @Override
+  public void validate(FormatContext context) {
+    delegate.validate(context);
+  }
+
+  @Override
+  public String getOutputFormatClassName() {
+    return GCSOutputFormat.class.getName();
+  }
+
+  @Override
+  public Map<String, String> getOutputFormatConfiguration() {
+    Map<String, String> outputFormatConfiguration = new HashMap<>(delegate.getOutputFormatConfiguration());
+    outputFormatConfiguration.put(DELEGATE_OUTPUTFORMAT_CLASSNAME, delegate.getOutputFormatClassName());
+    return outputFormatConfiguration;
+  }
+
+  /**
+   * OutputFormat for GCS Sink
+   */
+  public static class GCSOutputFormat extends OutputFormat<NullWritable, StructuredRecord> {
+    private OutputFormat delegateFormat;
+
+
+    private OutputFormat getDelegateFormatInstance(Configuration configuration) throws IOException {
+      String delegateClassName = configuration.get(DELEGATE_OUTPUTFORMAT_CLASSNAME);
+      if (delegateFormat == null) {
```

Comment:
nit: it's more readable to check the inverse, as it removes one layer of nesting:

```
if (delegateFormat != null) {
  return delegateFormat;
}
...
```

Reply:
Done.